kv_test.go

```diff
@@ -195,6 +196,123 @@ func TestCompareAndDelete(t *testing.T) {
 	assert.True(t, deleted)
 }
 
+func TestSetAtomicWithRetries(t *testing.T) {
+	tests := []struct {
+		name      string
+		key       string
+		valueFunc func(oldValue interface{}) (newValue interface{}, err error)
+		setupAPI  func(api *plugintest.API)
+		wantErr   bool
+	}{
+		{
+			name: "Test SetAtomicWithRetries success after first attempt",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
```

Comment:
Should we assert that the `old` value here is actually `oldJSONBytes` because it's possible that the func is passing in wrong `old` value and that these tests are still passing? thoughts?

Reply:
Ok, I added a test for that, but it's just testing whether the value is passed from line 168 to line 172--not much of a test. However, writing that test made me realize something I would have discovered as soon as I used this in the next PR: that if you pass an `interface{}` into get, it will marshal structs into a map, so `valueFunc` will have an annoying time coercing that back into the original type (map -> marshal -> []byte -> unmarshal -> expected type). I changed it to send `valueFunc` a `[]byte`. `valueFunc` can do what it wants with the []byte (convert it like I've done in the tests, or compare the raw `[]byte`, etc.), and return the `newValue` to be stored.

Reply:
I would argue what's being passed into the `valueFunc` is quite important for cases where we're mutating the `oldValue` and returning that. If the `oldValue` being passed in is invalid that may result in storing an invalid value.

Regarding passing in `interface{}` vs `[]byte`, is converting the only annoyance there? I wonder if we can write a helper to address that? As it is odd to see `valueFunc` getting `[]byte` but returning `interface{}` and that will lead to confusion. Thoughts?

Reply:
@cpoile this pr is good to go once the above concern is addressed:

> Regarding passing in interface{} vs []byte, is converting the only annoyance there? I wonder if we can write a helper to address that? As it is odd to see valueFunc getting []byte but returning interface{} and that will lead to confusion. Thoughts?

Reply:
@ali-farooq0 Yes, I agree it's a little strange to see `valueFunc` take a `[]byte` and return an `interface{}`, but only at first. It makes sense once you realize that the `interface{}` being passed in is not what you expect it to be. You can't cast it to your original type. That is a hidden gotcha that the users have to learn, and they would probably have to relearn it each time. When you pass in a `[]byte` at least it is clear what it is and that you have to interpret it yourself. Good?

Reply:
Ahh i see, i see. Do you what the underlying type is that's passed in when it was `interface{}`?

Reply:
By default, structs are marshaled into map[string]interface{}

Reply:
ooo.... gotcha well that's unfortunate :/ your approach makes sense in that case. The other option could be to use `json.RawMessage` but that's essentially the same thing so 0/5. Alright Cool!